webapp/app/routes/app/freeze/_route.tsx

```diff
@@ -0,0 +1,195 @@
+import { createFreezeProposal, getAllFreezeProposals } from "@/.server/db/dto/freeze-proposals";
+import { isUniqueConstraintError } from "@/.server/db/errors";
+import {
+  type FreezeProposalsType,
+  type freezeProposalsTable,
+  freezeProposalsTypeSchema,
+} from "@/.server/db/schema";
+import {
+  councilHardFreezeNonce,
+  councilSoftFreezeNonce,
+  councilSoftFreezeThresholdSettingNonce,
+  councilUnfreezeNonce,
+} from "@/.server/service/contracts";
+import { Badge } from "@/components/ui/badge";
+import { Button } from "@/components/ui/button";
+import { Card, CardContent, CardHeader, CardTitle } from "@/components/ui/card";
+import { CreateFreezeProposalModal } from "@/routes/app/freeze/create-freeze-proposal-modal";
+import { formError, generalError } from "@/utils/action-errors";
+import { cn } from "@/utils/cn";
+import type { ActionFunctionArgs } from "@remix-run/node";
+import { Link, json, redirect, useLoaderData } from "@remix-run/react";
+import type { InferSelectModel } from "drizzle-orm";
+import { ArrowRight, PlusIcon } from "lucide-react";
+import { useState } from "react";
+import { getFormData } from "remix-params-helper";
+import { $path } from "remix-routes";
+import type { Jsonify } from "type-fest";
+import { z } from "zod";
+
+export async function loader() {
+  const proposals = await getAllFreezeProposals();
+
+  // Filter expired proposals
+  const now = new Date();
+  const validProposals = proposals.filter((p) => new Date(p.validUntil) > now);
+
+  return json({
+    softFreezeProposals: validProposals.filter((p) => p.type === "SOFT_FREEZE"),
+    hardFreezeProposals: validProposals.filter((p) => p.type === "HARD_FREEZE"),
+    setSoftFreezeThresholdProposals: validProposals.filter(
+      (p) => p.type === "SET_SOFT_FREEZE_THRESHOLD"
+    ),
+    unfreezeProposals: validProposals.filter((p) => p.type === "UNFREEZE"),
+  });
+}
+
+export async function action({ request }: ActionFunctionArgs) {
+  const data = await getFormData(
+    request,
+    z
+      .object({
+        validUntil: z.coerce.date().min(new Date()),
+        threshold: z.number().min(1).max(9).optional(),
+        type: freezeProposalsTypeSchema,
+      })
+      .refine(
+        (data) => {
+          if (data.type === "SET_SOFT_FREEZE_THRESHOLD") {
+            return data.threshold !== undefined;
+          }
+          return data.threshold === undefined;
+        },
+        {
+          message: "Threshold must be defined only when type is SET_SOFT_FREEZE_THRESHOLD",
+          path: ["threshold"],
+        }
+      )
+  );
+  if (!data.success) {
+    return json(formError(data.errors));
+  }
+
+  let nonce: bigint;
+  switch (data.data.type) {
```

Comment:
This switch is replicated twice. I would love to make this a shared function.

A maybe nice way to implement this could be:

``` ts
// utis/freeze_threshold.ts
const GETTERS = {
SOFT_FREEZE:  councilSoftFreezeNonce,
HARD_FREEZE: councilHardFreezeNonce,
// ...
}

export async function freezeThreshold (type: Type): Promise<number> {
  return GETTERS[type]
}
```